compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
             } else {
                 format!("{}.dll", name)
             };
-            dylib_table
-                .entry(name)
-                .or_default()
-                .extend(lib.dll_imports.iter().map(|import| import.name));
+            dylib_table.entry(name).or_default().extend(lib.dll_imports.iter().cloned());
         }
     }
 
-    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
-    // what we should do if we have two DllImport values with the same name but different
-    // ordinals.
+    // Check for multiple imports with the same name but different calling conventions or
+    // (when relevant) argument list sizes.  Rustc only signals an error for this if the
+    // declarations are at the same scope level; if one shadows the other, we only get a lint
+    // warning.
+    for (library, imports) in &dylib_table {
+        let mut import_table: FxHashMap<Symbol, &DllImport> = FxHashMap::default();
+        for import in imports {
+            if let Some(old_import) = import_table.insert(import.name, &import) {
+                if import != old_import {
+                    sess.span_fatal(
+                        import.span,
+                        &format!(
+                            "multiple inconsistent definitions of external function `{}` from library `{}`",
```

Comment:
It's better to explicitly say that the functions have different ABIs, otherwise it's not clear how exactly they are inconsistent.